pkg/pillar/pubsub/pubsub.go

```diff
@@ -6,455 +6,68 @@
-			slaveCollection[masterKey] = deepCopy(master)
-			keys = append(keys, masterKey)
-		} else if !cmp.Equal(master, *slave) {
-			log.Debugf("determineDiffs(%s): key %s replacing due to diff %v\n",
-				name, masterKey,
-				cmp.Diff(master, *slave))
-			// XXX is deepCopy needed?
-			slaveCollection[masterKey] = deepCopy(master)
-			keys = append(keys, masterKey)
-		} else {
-			log.Debugf("determineDiffs(%s): key %s unchanged\n",
-				name, masterKey)
-		}
-	}
-	return keys
-}
-
-func SockName(name string) string {
-	return fmt.Sprintf("/var/run/%s.sock", name)
-}
-
-func PubDirName(name string) string {
-	return fmt.Sprintf("/var/run/%s", name)
-}
-
-func FixedDirName(name string) string {
-	return fmt.Sprintf("%s/%s", fixedDir, name)
-}
-
-func PersistentDirName(name string) string {
-	return fmt.Sprintf("%s/status/%s", "/persist", name)
-}
-
-// One shot create directory and publish one key in that directory
-func PublishToDir(dirName string, key string, item interface{}) error {
-	topic := TypeToName(item)
-	pub := new(Publication)
-	pub.topicType = item
-	pub.topic = topic
-	pub.km = keyMap{key: NewLockedStringMap()}
-	dirName = fmt.Sprintf("%s/%s", dirName, pub.topic)
-	pub.dirName = dirName
-	pub.publishToDir = true
-	name := pub.nameString()
-
-	if _, err := os.Stat(dirName); err != nil {
-		log.Infof("PublishToDir Create %s\n", dirName)
-		if err := os.MkdirAll(dirName, 0700); err != nil {
-			errStr := fmt.Sprintf("Publish(%s): %s",
-				name, err)
-			return errors.New(errStr)
-		}
-	}
-	return pub.Publish(key, item)
+// Usage:
+//  p1, err := pubsub.Publish("foo", fooStruct{})
+//  ...
+//  // Optional
+//  p1.SignalRestarted()
+//  ...
+//  p1.Publish(key, item)
+//  p1.Unpublish(key) to delete
+//
+//  foo := p1.Get(key)
+//  fooAll := p1.GetAll()
+
+type Publication interface {
+	Publish(key string, item interface{}) error
+	Unpublish(key string) error
+	SignalRestarted() error
+	ClearRestarted() error
+	Get(key string) (interface{}, error)
+	GetAll() map[string]interface{}
+}
+
+// SubHandler is a generic handler to handle create, modify and delete
+// Usage:
+//  s1 := pubsub.Subscribe("foo", fooStruct{}, true, &myctx)
+// Or
+//  s1 := pubsub.Subscribe("foo", fooStruct{}, false, &myctx)
+//  s1.ModifyHandler = func(...), // Optional
+//  s1.DeleteHandler = func(...), // Optional
+//  s1.RestartHandler = func(...), // Optional
+//  [ Initialize myctx ]
+//  s1.Activate()
+//  ...
+//  select {
+//     case change := <- s1.C:
+//         s1.ProcessChange(change, ctx)
+//  }
+type SubHandler func(ctx interface{}, key string, status interface{})
+type SubRestartHandler func(ctx interface{}, restarted bool)
+
+type Subscription interface {
+	Channel() <-chan string
+	Activate() error
+	ProcessChange(change string)
+	Get(key string) (interface{}, error)
+	GetAll() map[string]interface{}
+	Restarted() bool
+	Synchronized() bool
+	Handlers() (SubHandler, SubHandler, SubHandler, SubRestartHandler, SubRestartHandler)
```

Comment:
Why would we need this??

Reply:
There are lots of places where we do things like:

```go
// sub is a pubsub.Subscription
sub.CreateHandler = ...
sub.ModifyHandler = ...
```

If it is an interface, we need to define an API; you cannot change members directly.